R/utils.R

```diff
@@ -182,101 +203,112 @@ longwide <-
+    wt <- wt_id <- ht <- ht_id <- NULL
+
+    # check that all needed columns are present, if not, throw an error.
+    if(all(c(id, subjid, sex, agedays, param, measurement, gcr_result,
+             extra_cols) %in% colnames(long_df))) {
+      long_df %>%
+        select(id, subjid, sex, agedays, param, measurement, gcr_result,
+               all_of(extra_cols)) -> obs_df
+    } else {
+      # catch error if any variables were not found
+      stop("not all needed columns were present")
+    }
 
-  obs_df %>%
-    mutate(sex = sex_recoded) %>%
-    mutate(param = as.character(param)) %>%
-    select(subjid, id, agey, agem, agedays, sex, param, measurement) -> clean_df
+    # extract values flagged with indicated inclusion types:
+    if (include_all == TRUE) {
+      obs_df <- obs_df
+    } else if (include_all == FALSE) {
+      obs_df <- obs_df[obs_df[,gcr_result] %in% inclusion_types,]
+    } else{
+      stop(paste0("include_all is not a logical of length 1. It is a ",
+                  typeof(include_all), " of length ", length(include_all)))
+    }
 
+    # calculate age in years
+    obs_df$agey <- round(obs_df$agedays / 365.25, 4)
+
+    # calculate age in months
+    obs_df$agem <- round((obs_df$agey * 12), 4)
+
+    # recode sex to expected ext_bmiz() format
+    obs_df <- recode_sex(
+      input_data = obs_df,
+      sourcecol = "sex",
+      sourcem = "0",
+      sourcef = "1",
+      targetcol = "sex_recoded",
+      targetm = 1L,
+      targetf = 2L
+    )
+
+    obs_df %>%
+      mutate(sex = sex_recoded) %>%
+      mutate(param = as.character(param)) %>%
+      select(subjid, id, agey, agem, agedays, sex, param, measurement,
+             all_of(extra_cols)) -> clean_df
+
+
+    # check for unique weight and height ids
+    if (any(duplicated(clean_df$id))) {
+      stop("duplicate IDs in long_df")
+    }
 
-  # check for unique weight and height ids
-  if (any(duplicated(clean_df$id))) {
-    stop("duplicate IDs in long_df")
-  }
+    # separate heights and weights using unique ids
+    clean_df %>%
+      pivot_wider(names_from = param, values_from = measurement) -> param_separated
+
+    # extract heights and weights attached to ids
+    param_separated %>%
+      filter(!is.na(HEIGHTCM)) %>%
+      filter(is.na(WEIGHTKG)) %>% # why this, shouldn't the !is.na height take care of it
+      mutate(ht_id = id) %>%
+      select(-id) %>%
+      select(-WEIGHTKG) -> height
+
+    param_separated %>%
+      filter(is.na(HEIGHTCM)) %>%
+      filter(!is.na(WEIGHTKG)) %>%
+      mutate(wt_id = id) %>%
+      select(-id) %>%
+      select(-HEIGHTCM) -> weight
+
+    # join based on subjid, age, and sex, potentially keep unmatched values too
+    wide_df <- merge(height,
+                     weight,
+                     by = c("subjid", "agey", "agem", "agedays", "sex"),
+                     all = keep_unmatched_data) %>%
+      mutate(wt = WEIGHTKG, ht = HEIGHTCM)
+
+
+    for(c in extra_cols) {
```

Comment:
Don't use c -- that's a overwriting a function c()

Reply:
Sorry I missed that, change made.